README.md
RPowerFlow
==========

This project provides a steady-state electric power flow program written
in Ruby, based off both the Ruby linalg and Ruby narray libraries.

The program will provide both AC and DC solvers, and the AC solver will be
capable of both Full and Decoupled Newton solves.

The [Ruby linalg library](http://github.com/quix/linalg) is a Fortran-based
linear albegra library, and the [Ruby narray library](http://narray.rubyforge.org/)
is a pure-Ruby linear algebra library. Neither are capable of handling sparse
matrices, thus this project will have a limit as to how big of a power system
it can handle.

Installation
============

RPowerFlow can be installed via RubyGems by running the following command:

    $> gem install rpowerflow

This will install the narray gem as well, but will not install the linalg
library. See the "Requirements" section below for information about installing
the linalg library.

Current Status
==============

As of 04-21-2010, both the linalg and narray libraries are supported. Linalg is
loaded first, and if that fails then narray is loaded. A Full Newton solver is
implemented, and two tests exist.

Requirements
============

Either the Ruby linalg or Ruby narray package is necessary (just one, not both).

Ruby linalg is somewhat hard to install on later versions of Ubuntu since it
relies on some Fortran libraries no longer provided in the repositories. One
way around this is to download the required .deb file from the Debian Etch
repository and force the install.

Ruby narray can be installed as a gem.

TODO
====

* Check for expected methods on buses and branches as they are added to the system
* Add a Decoupled Newton solver
* Add a DC solver
* Provide examples of more complex systems
* Allow for enforcement of generator reactive power limits

Legal Notice
============

Copyright (2006-2010) Sandia Corporation.
Under the terms of Contract DE-AC04-94AL85000 with Sandia Corporation,
the U.S. Government retains certain rights in this software.

Contact: Bryan T. Richardson, Sandia National Laboratories, btricha@sandia.gov

This library is free software; you can redistribute it and/or modify it
under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation; either version 2.1 of the License, or (at
your option) any later version.

This library is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
details.

You should have received a copy of the GNU Lesser General Public License
along with this library; if not, write to the Free Software Foundation, Inc.,
59 Temple Place, Suite 330, Boston, MA 02111-1307 USA 


CHANGELOG
RPowerFlow CHANGELOG

VERSION 0.5.1
 * Removed reference to a RubyForge project from the Gemspec.
 * Included this CHANGELOG file in the gem.

VERSION 0.5.0
 * Initial version of project when provided via open source.


test/6_bus_system_test.rb
# Copyright (2006-2010) Sandia Corporation.
# Under the terms of Contract DE-AC04-94AL85000 with Sandia Corporation,
# the U.S. Government retains certain rights in this software.
#
# Author: Bryan T. Richardson, Sandia National Laboratories, btricha@sandia.gov
#
# This library is free software; you can redistribute it and/or modify it
# under the terms of the GNU Lesser General Public License as published by
# the Free Software Foundation; either version 2.1 of the License, or (at
# your option) any later version.
#
# This library is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
# details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with this library; if not, write to the Free Software Foundation, Inc.,
# 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA 

$LOAD_PATH.unshift(File.expand_path(File.dirname(__FILE__) + '/../lib'))

require 'rpowerflow'

Struct.new('Bus', :id, :voltage, :angle, :mw, :mvar, :slack, :avr, :susceptance)
Struct.new('Branch', :id, :resistance, :reactance, :charging)

system = ::RPowerFlow::System.new

bus01 = Struct::Bus.new(1, 1.0, 0.0,  0.00,  0.00,  true, true,  0.0)
bus02 = Struct::Bus.new(2, 1.0, 0.0,  1.10, -0.04, false, true,  0.0)
bus03 = Struct::Bus.new(3, 1.0, 0.0, -0.85, -0.17, false, false, 0.0)
bus04 = Struct::Bus.new(4, 1.0, 0.0, -0.40,  0.08, false, false, 0.0)
bus05 = Struct::Bus.new(5, 1.0, 0.0, -0.20, -0.04, false, false, 0.0)
bus06 = Struct::Bus.new(6, 1.0, 0.0, -0.20, -0.04, false, false, 0.0)

system << bus01
system << bus02
system << bus03
system << bus04
system << bus05
system << bus06

line01 = Struct::Branch.new(1, 0.0912, 0.48, 0.0282)
line02 = Struct::Branch.new(2, 0.0342, 0.18, 0.0106)
line03 = Struct::Branch.new(3, 0.1140, 0.60, 0.0352)
line04 = Struct::Branch.new(4, 0.0228, 0.12, 0.0071)
line05 = Struct::Branch.new(5, 0.0228, 0.12, 0.0071)
line06 = Struct::Branch.new(6, 0.0228, 0.12, 0.0071)
line07 = Struct::Branch.new(7, 0.0228, 0.12, 0.0071)
line08 = Struct::Branch.new(8, 0.0342, 0.18, 0.0106)
line09 = Struct::Branch.new(9, 0.1140, 0.60, 0.0352)

system[bus01, bus02] = line01
system[bus01, bus03] = line02
system[bus02, bus04] = line03
system[bus03, bus04] = line04
system[bus03, bus05] = line05
system[bus04, bus05] = line06
system[bus05, bus06] = line07
system[bus01, bus03] = line08
system[bus02, bus04] = line09

puts "Solved in #{::RPowerFlow::Solver::FullNewton.solve(system, 50, 0.01)} iterations"
puts

system.buses.each { |b| puts b.inspect }


test/14_bus_system_test.rb
# Copyright (2006-2010) Sandia Corporation.
# Under the terms of Contract DE-AC04-94AL85000 with Sandia Corporation,
# the U.S. Government retains certain rights in this software.
#
# Author: Bryan T. Richardson, Sandia National Laboratories, btricha@sandia.gov
#
# This library is free software; you can redistribute it and/or modify it
# under the terms of the GNU Lesser General Public License as published by
# the Free Software Foundation; either version 2.1 of the License, or (at
# your option) any later version.
#
# This library is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
# details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with this library; if not, write to the Free Software Foundation, Inc.,
# 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA 

$LOAD_PATH.unshift(File.expand_path(File.dirname(__FILE__) + '/../lib'))

require 'rpowerflow'

Struct.new('Bus', :id, :voltage, :angle, :mw, :mvar, :slack, :avr, :susceptance)
Struct.new('Branch', :id, :resistance, :reactance, :charging)

system = ::RPowerFlow::System.new

bus01 = Struct::Bus.new(1,  1.060, 0.0,  0.000,  0.000,  true, true,  0.0)
bus02 = Struct::Bus.new(2,  1.045, 0.0,  0.183, -0.127, false, true,  0.0)
bus03 = Struct::Bus.new(3,  1.010, 0.0, -0.942, -0.190, false, true,  0.0)
bus04 = Struct::Bus.new(4,  1.000, 0.0, -0.478,  0.039, false, false, 0.0)
bus05 = Struct::Bus.new(5,  1.000, 0.0, -0.076, -0.016, false, false, 0.0)
bus06 = Struct::Bus.new(6,  1.070, 0.0, -0.112, -0.075, false, true,  0.0)
bus07 = Struct::Bus.new(7,  1.000, 0.0,  0.000,  0.000, false, false, 0.0)
bus08 = Struct::Bus.new(8,  1.090, 0.0,  0.000,  0.000, false, true,  0.0)
bus09 = Struct::Bus.new(9,  1.000, 0.0, -0.295, -0.166, false, false, 0.0)
bus10 = Struct::Bus.new(10, 1.000, 0.0, -0.090, -0.058, false, false, 0.0)
bus11 = Struct::Bus.new(11, 1.000, 0.0, -0.035, -0.018, false, false, 0.0)
bus12 = Struct::Bus.new(12, 1.000, 0.0, -0.061, -0.016, false, false, 0.0)
bus13 = Struct::Bus.new(13, 1.000, 0.0, -0.135, -0.058, false, false, 0.0)
bus14 = Struct::Bus.new(14, 1.000, 0.0, -0.149, -0.050, false, false, 0.0)

system << bus01
system << bus02
system << bus03
system << bus04
system << bus05
system << bus06
system << bus07
system << bus08
system << bus09
system << bus10
system << bus11
system << bus12
system << bus13
system << bus14

line01 = Struct::Branch.new(1,  0.01938, 0.05917, 0.0528)
line02 = Struct::Branch.new(2,  0.05403, 0.22304, 0.0492)
line03 = Struct::Branch.new(3,  0.04699, 0.19797, 0.0438)
line04 = Struct::Branch.new(4,  0.05811, 0.17632, 0.0340)
line05 = Struct::Branch.new(5,  0.05695, 0.17388, 0.0346)
line06 = Struct::Branch.new(6,  0.06701, 0.17103, 0.0128)
line07 = Struct::Branch.new(7,  0.01335, 0.04211, 0.0000)
line08 = Struct::Branch.new(8,  0.00000, 0.20912, 0.0000)
line09 = Struct::Branch.new(9,  0.00000, 0.55618, 0.0000)
line10 = Struct::Branch.new(10, 0.00000, 0.25202, 0.0000)
line11 = Struct::Branch.new(11, 0.09498, 0.19890, 0.0000)
line12 = Struct::Branch.new(12, 0.12291, 0.25581, 0.0000)
line13 = Struct::Branch.new(13, 0.06615, 0.13027, 0.0000)
line14 = Struct::Branch.new(14, 0.00000, 0.17615, 0.0000)
line15 = Struct::Branch.new(15, 0.00000, 0.11001, 0.0000)
line16 = Struct::Branch.new(16, 0.03181, 0.08450, 0.0000)
line17 = Struct::Branch.new(17, 0.12711, 0.27038, 0.0000)
line18 = Struct::Branch.new(18, 0.08205, 0.19207, 0.0000)
line19 = Struct::Branch.new(19, 0.22092, 0.19988, 0.0000)
line20 = Struct::Branch.new(20, 0.17093, 0.34802, 0.0000)

system[bus01, bus02] = line01
system[bus01, bus05] = line02
system[bus02, bus03] = line03
system[bus02, bus04] = line04
system[bus02, bus05] = line05
system[bus03, bus04] = line06
system[bus04, bus05] = line07
system[bus04, bus07] = line08
system[bus04, bus09] = line09
system[bus05, bus06] = line10
system[bus06, bus11] = line11
system[bus06, bus12] = line12
system[bus06, bus13] = line13
system[bus07, bus08] = line14
system[bus07, bus09] = line15
system[bus09, bus10] = line16
system[bus09, bus14] = line17
system[bus10, bus11] = line18
system[bus12, bus13] = line19
system[bus13, bus14] = line20

puts "Solved in #{::RPowerFlow::Solver::FullNewton.solve(system, 100, 0.1)} iterations"
puts

system.buses.each { |b| puts b.inspect }


lib/rpowerflow.rb
# Copyright (2006-2010) Sandia Corporation.
# Under the terms of Contract DE-AC04-94AL85000 with Sandia Corporation,
# the U.S. Government retains certain rights in this software.
#
# Author: Bryan T. Richardson, Sandia National Laboratories, btricha@sandia.gov
#
# This library is free software; you can redistribute it and/or modify it
# under the terms of the GNU Lesser General Public License as published by
# the Free Software Foundation; either version 2.1 of the License, or (at
# your option) any later version.
#
# This library is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
# details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with this library; if not, write to the Free Software Foundation, Inc.,
# 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA 

require 'rpowerflow/solver'
require 'rpowerflow/system'


lib/rpowerflow/solver.rb
# Copyright (2006-2010) Sandia Corporation.
# Under the terms of Contract DE-AC04-94AL85000 with Sandia Corporation,
# the U.S. Government retains certain rights in this software.
#
# Author: Bryan T. Richardson, Sandia National Laboratories, btricha@sandia.gov
#
# This library is free software; you can redistribute it and/or modify it
# under the terms of the GNU Lesser General Public License as published by
# the Free Software Foundation; either version 2.1 of the License, or (at
# your option) any later version.
#
# This library is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
# details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with this library; if not, write to the Free Software Foundation, Inc.,
# 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA 

require 'rpowerflow/solvers/full_newton'


lib/rpowerflow/solvers/full_newton.rb
# Copyright (2006-2010) Sandia Corporation.
# Under the terms of Contract DE-AC04-94AL85000 with Sandia Corporation,
# the U.S. Government retains certain rights in this software.
#
# Author: Bryan T. Richardson, Sandia National Laboratories, btricha@sandia.gov
#
# This library is free software; you can redistribute it and/or modify it
# under the terms of the GNU Lesser General Public License as published by
# the Free Software Foundation; either version 2.1 of the License, or (at
# your option) any later version.
#
# This library is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
# details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with this library; if not, write to the Free Software Foundation, Inc.,
# 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA 

require 'complex'

begin
  require 'linalg'
rescue LoadError
  require 'rubygems'
  require 'narray'
end

module RPowerFlow
  module Solver
    class FullNewton
      # Currently returns the number of iterations required to
      # solve the system provided. Updates the system in place
      # with the calculated values.
      #
      # TODO: exceptions!
      # Throw an exception if the solver fails to converge.
      # Both the Linalg and NArray packages throw an exception
      # when an LU solve is attempted against a singular matrix.
      #
      # TODO: logging!
      def self.solve(system, max_iterations, tolerance)
        size       = system.buses.length
        iterations = 0

        max_iterations.times do
          # First half of power and delta arrays are
          # Mw values, second half are Mvar values.
          #
          # TODO: reset rather than create (possible?)
          if defined?(Linalg)
            power = Linalg::DMatrix.columns [Array.new(size * 2, 0.0)]
            delta = Linalg::DMatrix.columns [Array.new(size * 2, 0.0)]
          else
            power = NVector.float(size * 2)
            delta = NVector.float(size * 2)
          end

          jacobian = system.jacobian do |i, mw, mvar|
            bus  = system.buses[i]

            power[i]        += mw
            power[i + size] += mvar

            # As the calculated power of each bus is updated
            # above, the change in power for each bus is also
            # calculated. We could do this at the end, but it
            # would require another loop. Not sure which is
            # more expensive... doing these calculations over
            # and over again unnecessarily or doing another loop.
            # Looks like a good spot for a benchmark!
            #
            # TODO: benchmark!
            delta[i]        = bus.mw - power[i]
            delta[i + size] = bus.mvar + (bus.voltage ** 2 * bus.susceptance) - power[i + size]
          end

          tolerable = true
          (0...size).each do |i|
            bus = system.buses[i]

            # In terms of power values, Mw is a given for voltage
            # regulated, non-slack buses and Mw and Mvar are givens
            # for load buses.
            #
            # If this is a voltage regulated bus and it's not
            # the slack bus, only check to see if Mw is within
            # error range. Else, it's a load bus, so check to
            # see if both Mw and Mvar are within error range.
            if bus.avr
              unless bus.slack
                tolerable = false if delta[i].abs > tolerance
              end
            else
              tolerable = false if delta[i].abs > tolerance || delta[i + size] > tolerance
            end
          end

          # If given power values are within error range,
          # update the system buses with the calculated
          # power values and end the solve. Otherwise, solve
          # the next iteration of voltage and angle values.
          if tolerable
            (0...size).each do |i|
              bus = system.buses[i]

              bus.mw   = power[i]
              bus.mvar = power[i + size]
            end

            # Break out of the iterations loop
            # since we've solved the system.
            break
          else
            if defined?(Linalg)
              x = Linalg::DMatrix.solve(jacobian, delta)
            else
              x = jacobian.lu.solve(delta)
            end

            # Update system bus voltage and angle
            # values such that they can be used for
            # the next round of power and Jacobian
            # formulations.
            (0...size).each do |i|
              bus = system.buses[i]

              bus.voltage += x[i + size]
              bus.angle   += x[i]
            end

            # Update the iteration count since we
            # had to do another linear solve.
            iterations += 1
          end
        end

        return iterations
      end
    end
  end
end


lib/rpowerflow/system.rb
# Copyright (2006-2010) Sandia Corporation.
# Under the terms of Contract DE-AC04-94AL85000 with Sandia Corporation,
# the U.S. Government retains certain rights in this software.
#
# Author: Bryan T. Richardson, Sandia National Laboratories, btricha@sandia.gov
#
# This library is free software; you can redistribute it and/or modify it
# under the terms of the GNU Lesser General Public License as published by
# the Free Software Foundation; either version 2.1 of the License, or (at
# your option) any later version.
#
# This library is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
# details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with this library; if not, write to the Free Software Foundation, Inc.,
# 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA 

require 'complex'

begin
  require 'linalg'
rescue LoadError
  require 'rubygems'
  require 'narray'
end

module RPowerFlow
  # TODO: logging!
  class System
    def initialize(args = Hash.new)
      @buses    = Hash.new
      @branches = Hash.new
    end

    def <<(*args)
      args.flatten.each do |bus|
        @buses[bus] = Array.new
      end

      @bus_array = nil
      @size      = nil
    end

    def []=(*args)
      branch            = args.pop
      @branches[branch] = args.flatten

      @branch_array = nil
      @ybus         = nil
    end

    def size
      return @size ||= buses.length
    end

    # TODO: check for required accessors
    def buses(branch = nil)
      if branch
        return @branches[branch]
      else
        return @bus_array ||= @buses.keys.sort { |a,b| a.id <=> b.id }
      end
    end

    # TODO: check for required accessors
    def branches(bus = nil)
      if bus
        return @buses[bus]
      else
        return @branch_array ||= @branches.keys
      end
    end

    def ybus
      return @ybus unless @ybus.nil?

      @ybus = Array.new(size) { Array.new(size, 0.0) }

      # TODO: account for transformers
      branches.each do |branch|
        impedance  = Complex(branch.resistance, branch.reactance)
        charging   = Complex(0, branch.charging)
        mini       = Array.new(2) { Array.new }
        mini[0][0] =  (1 / impedance) + charging
        mini[0][1] = -(1 / impedance)
        mini[1][0] =  mini[0][1]
        mini[1][1] =  mini[0][0]

        endpoints = buses(branch)
        source    = buses.index(endpoints.first)
        target    = buses.index(endpoints.last)

        @ybus[source][source] += mini[0][0]
        @ybus[source][target] += mini[0][1]
        @ybus[target][source] += mini[1][0]
        @ybus[target][target] += mini[1][1]
      end

      return @ybus
    end

    def jacobian(print = false)
      if defined?(Linalg)
        jac = Linalg::DMatrix.new(size * 2, size * 2)
      else
        jac = NMatrix.float(size * 2, size * 2)
      end

      (0...size).each do |i|
        source = buses[i]

        (0...size).each do |j|
          target = buses[j]

          if block_given?
            mw   = source.voltage * ybus[i][j].abs * target.voltage * Math.cos(source.angle - target.angle - ybus[i][j].arg)
            mvar = source.voltage * ybus[i][j].abs * target.voltage * Math.sin(source.angle - target.angle - ybus[i][j].arg)

            # Yield the index of the current bus and
            # the incremental Mw and Mvar values.
            yield i, mw, mvar
          end

          if source.slack
            if i == j
              jac[i,i]               = 1e+10
              jac[i + size,i + size] = 1e+10
#               jac[i,i]               = 1.0
#               jac[i + size,i + size] = 1.0
            else
              jac[i,j]               = 0.0
              jac[i,j + size]        = 0.0
              jac[j,i]               = 0.0
              jac[j + size,i]        = 0.0
              jac[i + size,j]        = 0.0
              jac[i + size,j + size] = 0.0
              jac[j,i + size]        = 0.0
              jac[j + size,i + size] = 0.0
            end
          else
            if i == j
              (0...size).each do |k|
                jac[i,i] += ybus[i][k].abs * buses[k].voltage * Math.sin(source.angle - buses[k].angle - ybus[i][k].arg) unless i == k

                if source.avr
                  jac[i + size,i + size] = 1e+10
#                   jac[i + size,i + size] = 1.0
                else
                  jac[i,i + size]        += ybus[i][k].abs * buses[k].voltage * Math.cos(source.angle - buses[k].angle - ybus[i][k].arg)
                  jac[i + size,i]        += ybus[i][k].abs * buses[k].voltage * Math.cos(source.angle - buses[k].angle - ybus[i][k].arg) unless i == k
                  jac[i + size,i + size] += ybus[i][k].abs * buses[k].voltage * Math.sin(source.angle - buses[k].angle - ybus[i][k].arg)
                end
              end

              jac[i,i] *= -source.voltage

              unless source.avr
                jac[i,i + size]        +=  source.voltage * ybus[i][i].abs * Math.cos(ybus[i][i].arg)
                jac[i + size,i]        *=  source.voltage
                jac[i + size,i + size] += -source.voltage * ybus[i][i].abs * Math.sin(ybus[i][i].arg)
              end
            else
              unless target.slack
                jac[i,j] = source.voltage * ybus[i][j].abs * target.voltage * Math.sin(source.angle - target.angle - ybus[i][j].arg)

                if source.avr
                  jac[i,j + size] = source.voltage * ybus[i][j].abs * Math.cos(source.angle - target.angle - ybus[i][j].arg) unless target.avr
                else
                  unless target.avr
                    jac[i,j + size]        =  source.voltage * ybus[i][j].abs * Math.cos(source.angle - target.angle - ybus[i][j].arg)
                    jac[i + size,j]        = -source.voltage * ybus[i][j].abs * target.voltage * Math.cos(source.angle - target.angle - ybus[i][j].arg)
                    jac[i + size,j + size] =  source.voltage * ybus[i][j].abs * Math.sin(source.angle - target.angle - ybus[i][j].arg)
                  end
                end
              end
            end
          end
        end
      end

      return jac
    end

    def print_ybus
      (0...size).each do |i|
        format = Array.new
        data   = Array.new

        (0...size).each do |j|
          element = ybus[i][j]

          format << "%.1f+%.1fj"
          data   << element.real
          data   << element.imag
        end

        puts format.join("\t") % data
        puts
      end
    end

    def print_jacobian
      jac = jacobian

      (0...size * 2).each do |i|
        format = Array.new
        data   = Array.new

        (0...size * 2).each do |j|
          format << "%.2f"
          data   << jac[i,j]
        end

        puts format.join("\t") % data
        puts
      end
    end
  end
end
